src/internal/agent/hooks/proxy.go

```diff
@@ -0,0 +1,224 @@
+	if err != nil {
+		message.Warnf("Unable to read the npmToken file within the zarf-agent pod.")
+		npmToken = []byte(zarfState.GitServer.PushPassword)
+	}
+
+	director := func(req *http.Request) {
+		message.Debugf("Before Req %#v", req)
+		message.Debugf("Before Req URL %#v", req.URL)
+
+		// We add this so that we can use it to rewrite urls in the response if needed
+		req.Header.Add("X-Forwarded-Host", req.Host)
+
+		// We remove this so that go will encode and decode on our behalf (see https://pkg.go.dev/net/http#Transport DisableCompression)
+		req.Header.Del("Accept-Encoding")
+
+		var targetURL *url.URL
+
+		if strings.HasPrefix(req.UserAgent(), "git") {
+			req.SetBasicAuth(zarfState.GitServer.PushUsername, zarfState.GitServer.PushPassword)
+
+			if strings.HasPrefix(req.URL.Path, noTransform) {
+				targetURL = noTransformTarget(zarfState.GitServer.Address, req.URL.Path)
+			} else {
+				// TODO: (@WSTARR) Remove hardcoded https, this doesn't come through on scheme, but we could check it from req.TLS (though we only serve https right now anyway)
+				transformedURL, err := git.TransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+				message.Debugf("%#v", err)
+
+				targetURL, err = url.Parse(transformedURL)
+				message.Debugf("%#v", err)
+			}
+		} else {
+			req.SetBasicAuth(zarfState.GitServer.PushUsername, zarfState.GitServer.PushPassword)
+
+			if strings.HasPrefix(req.URL.Path, noTransform) {
+				targetURL = noTransformTarget(zarfState.GitServer.Address, req.URL.Path)
+			} else {
+				transformedURL := ""
+
+				if strings.HasPrefix(req.UserAgent(), "pip") || strings.HasPrefix(req.UserAgent(), "twine") {
+					transformedURL = pipTransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+					// TODO: (@WSTARR) We will need to support separate git repos and package registries in the future
+				} else if strings.HasPrefix(req.UserAgent(), "npm") || strings.HasPrefix(req.UserAgent(), "yarn") {
+					transformedURL = npmTransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+					// TODO: (@WSTARR) We will need to support separate external git repos and package registries in the future
+					req.Header.Set("Authorization", "Bearer "+string(npmToken))
+				} else {
+					transformedURL = genTransformURL(zarfState.GitServer.Address, "https://"+req.Host+req.URL.String(), zarfState.GitServer.PushUsername)
+					// TODO: (@WSTARR) We will need to support separate external git repos and package registries in the future
+				}
+
+				targetURL, err = url.Parse(transformedURL)
+				message.Debugf("%#v", err)
+			}
+		}
+
+		req.Host = targetURL.Host
+		req.URL = targetURL
+
+		message.Debugf("After Req %#v", req)
+		message.Debugf("After Req URL%#v", req.URL)
+	}
+
+	response := func(resp *http.Response) error {
```

Comment:
nit; can we define this response func somewhere else and just assign it here? I think that'll make it easier to read the flow of the current function that way.

Reply:
I did a big refactor and moved everything around now.